<launch>
  <arg name="rviz" default="false" />
  <arg name="publish_marker_image" default="true" />

  <arg name="cam1" default="cam1" />
  <arg name="cam2" default="cam2" />
  <arg name="cam3" default="cam3" />

  <arg name="cam1_video" default="/dev/video0" />
  <arg name="cam2_video" default="/dev/video4" />

   <group if="$(arg rviz)">
     <node pkg="rviz" type="rviz" name="rviz_april" args="-d $(find state_estimation)/config/apriltag.rviz"  />
   </group>

  <!-- Set parameters -->
  <rosparam command="load" file="$(find state_estimation)/config/AT_settings.yaml" ns="apriltag_detector_1" />
  <rosparam command="load" file="$(find state_estimation)/config/AT_settings.yaml" ns="apriltag_detector_2" />
  <rosparam command="load" file="$(find state_estimation)/config/AT_tags.yaml" ns="apriltag_detector_1" />
  <rosparam command="load" file="$(find state_estimation)/config/AT_tags.yaml" ns="apriltag_detector_2" />
  
  <!--
    TAG DETECTOR 1
  -->
  <node pkg="usb_cam" type="usb_cam_node" name="usb_cam_$(arg cam1)" output="screen" >
        <param name="camera_name" value="$(arg cam1)" />
        <param name="pixel_format" value="yuyv" />
        <param name="video_device" value="$(arg cam1_video)" />
    </node>

  <node pkg="apriltag_ros" type="apriltag_ros_continuous_node" name="apriltag_detector_1" output="screen" >
    <remap from="image_rect" to="/usb_cam_$(arg cam1)/image_raw" />
    <remap from="camera_info" to="/usb_cam_$(arg cam1)/camera_info" />
    <remap from="tag_detections" to="/tag_detections_$(arg cam1)" />
    <remap from="tag_detections_image" to="/tag_detections_image_$(arg cam1)" />

    <param name="publish_tf" value="false" />
    <param name="publish_tag_detections_image" type="bool" value="$(arg publish_marker_image)" />
  </node>

  <!--
    TAG DETECTOR 2
  -->   
    <node pkg="usb_cam" type="usb_cam_node" name="usb_cam_$(arg cam2)" output="screen" >
        <param name="camera_name" value="$(arg cam2)" />
        <param name="pixel_format" value="yuyv" />
        <param name="video_device" value="$(arg cam2_video)" />
    </node>
    
  <node pkg="apriltag_ros" type="apriltag_ros_continuous_node" name="apriltag_detector_2" output="screen" >
    <!-- Remap topics from those used in code to those on the ROS network -->
    <remap from="image_rect" to="/usb_cam_$(arg cam2)/image_raw" />
    <remap from="camera_info" to="/usb_cam_$(arg cam2)/camera_info" />
    <remap from="tag_detections" to="/tag_detections_$(arg cam2)" />
    <remap from="tag_detections_image" to="/tag_detections_image_$(arg cam2)" />

    <param name="publish_tf" value="false" />
    <param name="publish_tag_detections_image" type="bool" value="$(arg publish_marker_image)" />      <!-- default: false -->
  </node>

</launch>